Document datehistogram with long offsets (#93328)
elastic/elasticsearch: docs/reference/aggregations/bucket/datehistogram-aggregation.asciidoc

adjustments have been made.
// end::offset-note[]

[[search-aggregations-bucket-datehistogram-offset-months]]
===== Long offsets over calendar intervals

> As seen above, it is typical to use offsets in units smaller than the `calendar_interval`, for example

Reviewer: ```suggestion
It is typical to use offsets in units smaller than the `calendar_interval`. For example,
```

using offsets in hours when the interval is days, or an offset of days when the interval is months.
If the calendar interval is always of a standard length, or the `offset` is less than one unit of the calendar
interval (for example less than `+24h` for `days` or less than `+28d` for months),
then each bucket will have a repeating start. For example `+6h` for `days` will result in all buckets
starting at 6am each day. However, `+30h` will also result in buckets starting at 6am, except when crossing
days that change from standard to summer-savings time or vice-versa.

This situation is much more pronounced for months, where each month has a different length
to at least one of its adjacent months.
To demonstrate this, consider eight documents each with a date field on the 20th day of each of the
eight months from January to August of 2022.

> When querying for a date histogram over the calendar interval of months, we will have each document in a single bucket,
> one for each month.
> And each bucket will have a key named after the first day of the month, plus any offset.
> For example, the offset of `+19d` will have buckets with names like `2022-01-20`.

Reviewer: ```suggestion
When querying for a date histogram over the calendar interval of months, the response will return one bucket per month, each with a single document.
Each bucket will have a key named after the first day of the month, plus any offset.
For example, the offset of `+19d` will result in buckets with names like `2022-01-20`.
```

[source,console,id=datehistogram-aggregation-offset-example-19d]
--------------------------------------------------
"buckets": [
  { "key_as_string": "2022-01-20", "key": 1642636800000, "doc_count": 1 },
  { "key_as_string": "2022-02-20", "key": 1645315200000, "doc_count": 1 },
  { "key_as_string": "2022-03-20", "key": 1647734400000, "doc_count": 1 },
  { "key_as_string": "2022-04-20", "key": 1650412800000, "doc_count": 1 },
  { "key_as_string": "2022-05-20", "key": 1653004800000, "doc_count": 1 },
  { "key_as_string": "2022-06-20", "key": 1655683200000, "doc_count": 1 },
  { "key_as_string": "2022-07-20", "key": 1658275200000, "doc_count": 1 },
  { "key_as_string": "2022-08-20", "key": 1660953600000, "doc_count": 1 }
]
--------------------------------------------------
// TESTRESPONSE[skip:no setup made for this example yet]

> If we increase the offset to `+20d`, each document will appear in a bucket for the previous month,
> with all bucket keys ending with the same day of the month, as normal.
> However, if we further increase to `+28d`,
> we will find that what used to be a February bucket has now become `"2022-03-01"`.

Reviewer: ```suggestion
Increasing the offset to `+20d`, each document will appear in a bucket for the previous month,
with all bucket keys ending with the same day of the month, as normal.
However, further increasing to `+28d`, 
what used to be a February bucket has now become `"2022-03-01"`.
```

Documents that were originally 30 days apart can be shifted into the same 31-day month bucket.

For example, for `+50d` we see:

[source,console,id=datehistogram-aggregation-offset-example-50d]
--------------------------------------------------
"buckets": [
  { "key_as_string": "2022-01-20", "key": 1642636800000, "doc_count": 1 },
  { "key_as_string": "2022-02-20", "key": 1645315200000, "doc_count": 2 },
  { "key_as_string": "2022-04-20", "key": 1650412800000, "doc_count": 2 },
  { "key_as_string": "2022-06-20", "key": 1655683200000, "doc_count": 2 },
  { "key_as_string": "2022-08-20", "key": 1660953600000, "doc_count": 1 }
]
--------------------------------------------------
// TESTRESPONSE[skip:no setup made for this example yet]

It is therefor always important when using `offset` with `calendar_interval` bucket sizes
to understand the consequences of using offsets larger than the interval size.

More examples:

* If the goal is to, for example, have an annual histogram where each year starts on the 5th February,
you could use `calendar_interval` of `year` and `offset` of `+33d`, and each year will be shifted identically,

> because the offset includes only January, which ius the same length every year.

Reviewer: ```suggestion
because the offset includes only January, which is the same length every year.
```
